clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -459,7 +467,506 @@ void SemaHLSL::handleResourceClassAttr(Decl *D, const ParsedAttr &AL) {
+            << regType;
+      break;
+    case llvm::hlsl::ResourceClass::CBuffer:
+      if (regType != RegisterType::CBuffer)
+        S.Diag(TheDecl->getLocation(), diag::err_hlsl_binding_type_mismatch)
+            << regType;
+      break;
+    case llvm::hlsl::ResourceClass::Sampler:
+      if (regType != RegisterType::Sampler)
+        S.Diag(TheDecl->getLocation(), diag::err_hlsl_binding_type_mismatch)
+            << regType;
+      break;
+    }
+    return;
+  }
+
+  // next, handle diagnostics for when the "basic" flag is set,
+  // including the legacy "i" and "b" register types.
+  if (Flags.Basic) {
+    if (Flags.DefaultGlobals) {
+      if (regType == RegisterType::CBuffer)
+        S.Diag(ArgLoc, diag::warn_hlsl_deprecated_register_type_b);
+      else if (regType == RegisterType::I)
+        S.Diag(ArgLoc, diag::warn_hlsl_deprecated_register_type_i);
+      else if (regType != RegisterType::C)
+        S.Diag(ArgLoc, diag::err_hlsl_binding_type_mismatch) << regType;
+      return;
+    }
+
+    if (regType == RegisterType::C)
+      S.Diag(ArgLoc, diag::warn_hlsl_register_type_c_packoffset);
+    else if (regType == RegisterType::I)
+      S.Diag(ArgLoc, diag::warn_hlsl_deprecated_register_type_i);
+    else
+      S.Diag(ArgLoc, diag::err_hlsl_binding_type_mismatch) << regType;
+
+    return;
+  }
+
+  // finally, we handle the udt case
+  if (Flags.UDT) {
+    if (regType == RegisterType::I) {
+      S.Diag(TheDecl->getLocation(),
+             diag::warn_hlsl_deprecated_register_type_i);
+      return;
+    }
+    switch (getRegisterTypeIndex(Slot)) {
+    case RegisterType::SRV: {
+      if (!Flags.SRV) {
+        S.Diag(TheDecl->getLocation(),
+               diag::warn_hlsl_user_defined_type_missing_member)
+            << regType;
+      }
+      break;
+    }
+    case RegisterType::UAV: {
+      if (!Flags.UAV) {
+        S.Diag(TheDecl->getLocation(),
+               diag::warn_hlsl_user_defined_type_missing_member)
+            << getHLSLResourceTypeStr(S, TheDecl) << regType;
+      }
+      break;
+    }
+    case RegisterType::CBuffer: {
+      if (!Flags.CBV) {
+        S.Diag(TheDecl->getLocation(),
+               diag::warn_hlsl_user_defined_type_missing_member)
+            << getHLSLResourceTypeStr(S, TheDecl) << regType;
+      }
+      break;
+    }
+    case RegisterType::Sampler: {
+      if (!Flags.Sampler) {
+        S.Diag(TheDecl->getLocation(),
+               diag::warn_hlsl_user_defined_type_missing_member)
+            << getHLSLResourceTypeStr(S, TheDecl) << regType;
+      }
+      break;
+    }
+    case RegisterType::C: {
+      if (!Flags.ContainsNumeric)
+        S.Diag(TheDecl->getLocation(),
+               diag::warn_hlsl_user_defined_type_missing_member)
+            << regType;
+      break;
+    }
+    case RegisterType::I: {
+      S.Diag(TheDecl->getLocation(),
+             diag::warn_hlsl_deprecated_register_type_i);
+    }
+    default:
```

Comment:
This could probably be simplified with less repetition as well - the general pattern seems to be "for this register type, check this flag is true".